Minor speed boost and code formatting

AeroX2 opened the pull request:
I've formatted the code a little bit, so that the formatting is a little bit more consistent.

Merged the x9...x12 flag changes into this branch and separated the random start points into a separate kernel, this gave a small but noticeable boost in performance.

AeroX2 committed:
Merge pull request #1 from pack-png-mods/master

Merge latest changes from master

Commit:
Code rearrangement/formatting and small speed boost

Earthcomputer commented on main.cu:
```diff
@@ -77,27 +75,17 @@ __host__ __device__ inline uint random_next_int(Random *random, uint bound) {
-        for (int u = r;
-             u - (r = u % bound) + m < 0;
-             u = random_next(random, 31));
+        for (int u = r; u - (r = u % bound) + m < 0; u = random_next(random, 31));
```
No, please change this back. It was clearer on multiple lines IMO.

Earthcomputer commented on main.cu:
```diff
@@ -174,144 +133,187 @@ __device__ inline int getTreeHeight(int x, int z) {
+#define TREE_Z (WATERFALL_Z - 8)
+#define TREE_HEIGHT 5
+
+#define OTHER_TREE_COUNT 2
```
Please leave in the constants on the master branch, they are for a test seed for us to verify the correctness of the kernel.
Also, I tested the if statements in comparison to the tree flags, and the if statements were faster. So I only used the flags when I had to (wherever the exact location of a tree was not known). Does your testing have different results?

AeroX2 replied:
I'll have to do a more through analysis tomorrow, profiler is playing up.
I mainly added the flag system for consistency, seeing as I thought it would be merged with master, if I get the profiler working again and it is faster, I'll submit another PR to add the flags in.

Earthcomputer commented on main.cu:
```diff
@@ -174,144 +133,187 @@ __device__ inline int getTreeHeight(int x, int z) {
+        Random rand = rand_starts[i];
+        Random start = rand;
+
+        // Advance 10
```
This line doesn't advance by 10, it advances by 1 and returns a value from 0-9. This line can be omitted by changing the previous step to put the values of `rand` in the array, rather than `start`. Then, when adding to the seed list, advance by -3760 rather than -3759.

AeroX2 replied:
Hmmm, I did think of this however of was unsure about the mechanics of advance and if you could only advance by that constant.

That being said how are the values for advance calculated, I would need the new values for m3760.

Earthcomputer commented on main.cu:
```diff
@@ -174,144 +133,187 @@ __device__ inline int getTreeHeight(int x, int z) {
-                int index = atomicAdd(num_seeds, 1);
-                seeds[index] = start_chunk_rand;
-            }
+        advance_m1(start);
```
I'm not sure there's any need to advance start by -1 here, as it goes out of scope immediately afterwards.

AeroX2 replied:
Good catch.

Earthcomputer commented on main.cu:
```diff
@@ -350,64 +352,66 @@ void calculate_search_backs(int GPU_COUNT) {
+                break;
+            default:
+                printf("Error: Flag not recognized.");
+                return -1;
```
No, please indent case labels from the surrounding switch statement.

Earthcomputer commented on main.cu:
```diff
@@ -350,64 +352,66 @@ void calculate_search_backs(int GPU_COUNT) {
         }
     }
-    GPU_Node *nodes = (GPU_Node*)malloc(sizeof(GPU_Node) * GPU_COUNT);
+    GPU_Node *nodes = (GPU_Node *)malloc(sizeof(GPU_Node) * GPU_COUNT);
```
No, I don't like this formatting, please change it back.

Earthcomputer reviewed (changes requested):
Needs some changes before merge. In general, I hold the view that code formatting is not an automatable task, as sometimes to make specific code clear, it's okay to go against consistency. Most of the formatting you've done is okay, I have pointed out the ones which are not in my opinion.

Commit:
Updated code to comply with Github comments

Earthcomputer committed:
Add constants for advance_m3760

Earthcomputer reviewed (changes requested):
I pushed to your fork and added the constants. However, it is producing seeds that are about 100-106 random calls ahead of the old version.

AeroX2 commented:
Thanks for the constants, however I seem to get different results even on the original branch.
Modifying the original by adding the constants, changing start_chunk_rand = start to start_chunk_rand = rand, gives me different results than what the original branch produces but matches my modified branch.

Original:
108627439313040
112706455179245
117893381432493
118574646168622
140938423315907
142323255175552
143199981418189
154036868262394
158035617998632
166812489202574

Modified original:
111884084901320
121964594163720
123575461804449
139245505747986
143302516248640
16039266542546
172395983390754
20111672765370
206823937264321
214614257957367

Modified branch
111884084901320
123575461804449
139245505747986
143302516248640
16039266542546
172395983390754
20111672765370
206823937264321
214614257957367
216760090008747

```
james@jamespc ~/d/g/seed-reversal-original $ git diff           
diff --git a/main.cu b/main.cu
index 07cd536..152377b 100644
--- a/main.cu
+++ b/main.cu
@@ -105,7 +105,7 @@ inline void gpuAssert(cudaError_t code, const char* file, int line) {
 #define advance_16(rand) advance(rand, 0x6DC260740241LL, 0xD0352014D90LL)
 #define advance_m1(rand) advance(rand, 0xDFE05BCB1365LL, 0x615C0E462AA9LL)
 #define advance_m3759(rand) advance(rand, 0x63A9985BE4ADLL, 0xA9AA8DA9BC9BLL)
-
+#define advance_m3760(rand) advance(rand, 0x2DBFC2690F41LL, 0xB26D332614D0LL)
 
 
 #define TREE_X 4
@@ -283,8 +283,8 @@ __global__ void doWork(int* num_starts, Random* tree_starts, int* num_seeds, ulo
             this_res &= any_population_matches;
 
             if (this_res) {
-                Random start_chunk_rand = start;
-                advance_m3759(start_chunk_rand);
+                Random start_chunk_rand = rand;
+                advance_m3760(start_chunk_rand);
 
                 int index = atomicAdd(num_seeds, 1);
                 seeds[index] = start_chunk_rand;
james@jamespc ~/d/g/seed-reversal-original $ cat chunk_seeds.txt|sort|head
111884084901320
121964594163720
123575461804449
139245505747986
143302516248640
16039266542546
172395983390754
20111672765370
206823937264321
214614257957367
james@jamespc ~/d/g/seed-reversal-original $ cat ../seed-reversal-modified/chunk_seeds.txt|sort|head
111884084901320
123575461804449
139245505747986
143302516248640
16039266542546
172395983390754
20111672765370
206823937264321
214614257957367
216760090008747
